op-dispute-mon/mon/extract/output_agreement_enricher.go

```diff
@@ -30,54 +32,155 @@ type OutputMetrics interface {
+func NewOutputAgreementEnricher(logger log.Logger, metrics OutputMetrics, clients []OutputRollupClient, cl clock.Clock) *OutputAgreementEnricher {
 	return &OutputAgreementEnricher{
 		log:     logger,
 		metrics: metrics,
-		client:  client,
+		clients: clients,
 		clock:   cl,
 	}
 }
 
+type outputResult struct {
+	outputRoot common.Hash
+	isSafe     bool
+	notFound   bool
+	err        error
+}
+
 // Enrich validates the specified root claim against the output at the given block number.
 func (o *OutputAgreementEnricher) Enrich(ctx context.Context, block rpcblock.Block, caller GameCaller, game *monTypes.EnrichedGameData) error {
 	if !game.UsesOutputRoots() {
 		return nil
 	}
-	if o.client == nil {
+	if len(o.clients) == 0 {
 		return fmt.Errorf("%w but required for game type %v", ErrRollupRpcRequired, game.GameType)
 	}
-	output, err := o.client.OutputAtBlock(ctx, game.L2BlockNumber)
-	if err != nil {
-		// string match as the error comes from the remote server so we can't use Errors.Is sadly.
-		if strings.Contains(err.Error(), "not found") {
-			// Output root doesn't exist, so we must disagree with it.
-			game.AgreeWithClaim = false
-			return nil
+
+	results := make([]outputResult, len(o.clients))
+	var wg sync.WaitGroup
+	for i, client := range o.clients {
+		wg.Add(1)
+		go func(i int, client OutputRollupClient) {
+			defer wg.Done()
+			output, err := client.OutputAtBlock(ctx, game.L2BlockNumber)
+			if err != nil {
+				// string match as the error comes from the remote server so we can't use Errors.Is sadly.
+				if strings.Contains(err.Error(), "not found") {
+					results[i] = outputResult{notFound: true}
+					return
+				}
+				results[i] = outputResult{err: err}
+				return
+			}
+
+			outputRoot := common.Hash(output.OutputRoot)
+			results[i] = outputResult{outputRoot: outputRoot}
+
+			// Only check if the output root is safe if it matches the game's root claim
+			if outputRoot == game.RootClaim {
+				safeHead, err := client.SafeHeadAtL1Block(ctx, game.L1HeadNum)
+				if err != nil {
+					o.log.Warn("Unable to verify proposed block was safe", "l1HeadNum", game.L1HeadNum, "l2BlockNum", game.L2BlockNumber, "err", err)
+					// If safe head data isn't available, assume the output root was safe
+					// Avoids making the dispute mon dependent on safe head db being available
+					results[i].isSafe = true
+					return
+				}
+				results[i].isSafe = safeHead.SafeHead.Number >= game.L2BlockNumber
+			}
+		}(i, client)
+	}
+	wg.Wait()
+
+	validResults := make([]outputResult, 0, len(results))
+	foundResults := make([]outputResult, 0, len(results))
+	for idx, result := range results {
+		if result.err != nil {
+			o.log.Error("Failed to fetch output root", "clientIndex", idx, "l2BlockNum", game.L2BlockNumber, "err", result.err)
+			continue
+		}
+
+		validResults = append(validResults, result)
+
+		if result.notFound {
+			o.log.Info("Node has not seen block", "clientIndex", idx, "l2BlockNum", game.L2BlockNumber)
```

Comment:
I suspect we don't want to log this at info level because it will occur pretty much constantly so every update loop will wind up logging this multiple times (once for each game created by the invalid proposal since they all have a far future block number).